utils/estimated_histogram.hh

```diff
@@ -47,13 +47,376 @@
+        res.sample_sum += buckets[NUM_BUCKETS - 1] * get_bucket_limit(NUM_BUCKETS - 1);
+        return res;
+    }
+
+    /*!
+     * \brief return the histogram buckets
+     */
+    const std::vector<uint64_t>& get_buckets() const {
+        return buckets;
+    }
+
+    void clear() {
+        std::fill(buckets.begin(), buckets.end(), 0);
+    }
+
+    /*!
+     * \brief Add an item to the histogram
+     * Increments the count of the bucket closest to n, rounding DOWN.
+     */
+    histogram_type& add(uint64_t n) {
+        buckets.at(find_bucket_pos(n))++;
+        return *this;
+    }
+
+    /*!
+     * \brief returns the smallest value that could have been added to this histogram
+     */
+    uint64_t min() const {
+        for (size_t i = 0; i < NUM_BUCKETS; i ++) {
+            if (buckets[i] > 0) {
+                return get_bucket_limit(i);
+            }
+        }
+        return 0;
+    }
+
+    /*!
+     * \brief returns the largest value that could have been added to this histogram.  If the histogram
+     * overflowed, returns UINT64_MAX.
+     */
+    uint64_t max() const {
+        if (buckets[NUM_BUCKETS - 1] > 0) {
+            return UINT64_MAX;
+        }
+        for (int i = NUM_BUCKETS - 1; i >= 0; i--) {
+            if (buckets[i] > 0) {
+                return get_bucket_limit(i);
+            }
+        }
+        return 0;
+    }
+
+    /*!
+     * \brief merge a histogram to the current one.
+     */
+    histogram_type& merge(const histogram_type& b) {
+        for (size_t i = 0; i < NUM_BUCKETS; i++) {
+            buckets[i] += b.get_buckets()[i];
+        }
+        return *this;
+    }
+
+    template<uint64_t A, uint64_t B, std::vector<uint64_t>::size_type C>
+    friend pseudo_floatingpoint_histogram<A, B, C> merge(pseudo_floatingpoint_histogram<A, B, C> a, const pseudo_floatingpoint_histogram<A, B, C>& b);
+
+    /*
+     * \brief returns the count in the given bucket
+     */
+    uint64_t get(size_t bucket) const {
+        return buckets[bucket];
+    }
+
+    /*!
+     * \brief get a histogram percentile
+     *
+     * returns the estimated value at given percentile
+     */
+    int64_t percentile(double perc) const {
+        assert(perc >= 0 && perc <= 1.0);
+        auto last_bucket = NUM_BUCKETS - 1;
+
+        auto c = count();
+
+        if (!c) {
+            return 0; // no data
+        }
+
+        auto pcount = int64_t(std::floor(c * perc));
+        int64_t elements = 0;
+        for (size_t i = 0; i < last_bucket; i++) {
+            if (buckets[i]) {
+                elements += buckets[i];
+                if (elements >= pcount) {
+                    return get_bucket_limit(i);
+                }
+            }
+        }
+        return MAX; // overflowed value is in the requested percentile
+    }
+
+    /*!
+     * \brief return the mean histogram value (average of bucket offsets, weighted by count)
+     */
+    uint64_t mean() const {
```

Comment:
Again maybe return both bounds.

Reply:
The mean, and percentile are for internal usage and I wanted to keep the current API so it will be easy to migrate.